addon/devtoolsPanel.js

```diff
@@ -1,47 +1,232 @@
+        });
+        updateLabeled();
+
+    } else if (request.type === 'refresh') {
+        resetFreeze();
+        browser.runtime.sendMessage({type: 'init'})
+            .then(updateLabeled)
+            .catch((error) => {
+                console.error(error);
+            });
     }
+});
+
+// Restore the freeze button and spinner back to original state.
+function resetFreeze() {
+    document.getElementById('freeze-button').disabled = false;
+    document.getElementById('spinner-container').classList.add('hidden');
 }
 
-async function labelInspectedElement() {
-    /**
-     * The function embedded herein returns an "index path" to the inspected
-     * element. A return value like [1, 4, 0] means the element could be found by
-     * going into the 0th element of the document (usually an <html> tag), then the
-     * 4th child of that tag, then finding the 1st child of that tag.
-     */
-    backgroundPort.postMessage({type: 'label',
-                                tabId: browser.devtools.inspectedWindow.tabId,
-                                inspectedElement: await inspectedElementPath(),
-                                label: document.getElementById('labelField').value});
+// Returns a list of objects with the following attributes:
+//   preview: html preview of element
+//   label: the fathom label, or an empty string if unlabeled
+//   path: css path to element
+//   inspected: true if this element === $0
+function getLabeled() {
+    let evalScript = `
+        (function getLabeled() {
+            function elementMeta(el, isInspected) {
+                return {
+                    preview: el.outerHTML.replace(/^([^>]+>)[\\s\\S]*$/, '$1'),
+                    label: el.dataset.fathom || '',
+                    path: Simmer(el),
+                    inspected: isInspected,
+                };
+            }
+
+            let list = [];
+            if ($0) {
+                list.push(elementMeta($0, true));
+            }
+            for (const el of document.querySelectorAll('[data-fathom]')) {
+                if ($0 && el === $0) {
+                    continue;
+                }
+                list.push(elementMeta(el, false));
+            }
+            return list;
+        })();
+    `;
+    return resultOfEval(evalScript);
 }
-document.getElementById('labelButton').addEventListener('click', labelInspectedElement);
 
-async function freezePage() {
-    backgroundPort.postMessage({type: 'freeze',
-                                tabId: browser.devtools.inspectedWindow.tabId,
-                                inspectedElement: await inspectedElementPath(),
-                                options: {shouldScroll: false, wait: 0}});
+// Draw the table with the list of existing labeled elements.
+// If an element is inspected/selected with devtools (i.e. $0 is defined), it
+// will always be the first row in the table.
+function updateLabeledTable(labeled) {
```

Comment:
I didn't have a problem since I'm reading this in a straight line, but our future selves might enjoy a pointer to where the structure of `labeled` is documented.